libs/community/langchain_community/chat_models/outlines.py

```diff
@@ -0,0 +1,564 @@
+            if isinstance(tool_choice, dict):
+                if not any(
+                    tool_choice["function"]["name"] == name for name in tool_names
+                ):
+                    raise ValueError(
+                        f"Tool choice {tool_choice=} was specified, but the only "
+                        f"provided tools were {tool_names}."
+                    )
+            elif isinstance(tool_choice, str):
+                chosen = [
+                    f for f in formatted_tools if f["function"]["name"] == tool_choice
+                ]
+                if not chosen:
+                    raise ValueError(
+                        f"Tool choice {tool_choice=} was specified, but the only "
+                        f"provided tools were {tool_names}."
+                    )
+            elif isinstance(tool_choice, bool):
+                if len(formatted_tools) > 1:
+                    raise ValueError(
+                        "tool_choice=True can only be specified when a single tool is "
+                        f"passed in. Received {len(tools)} tools."
+                    )
+                tool_choice = formatted_tools[0]
+
+        kwargs["tool_choice"] = tool_choice
+        formatted_tools = [convert_to_openai_tool(tool) for tool in tools]
+        return super().bind_tools(tools=formatted_tools, **kwargs)
+
+    def with_structured_output(
+        self,
+        schema: Dict | type,
+        *,
+        include_raw: bool = False,
+        **kwargs: Any,
+    ) -> Runnable[
+        Union[
+            PromptValue,
+            str,
+            Sequence[
+                Union[BaseMessage, list[str], tuple[str, str], str, dict[str, Any]]
+            ],
+        ],
+        Union[dict, BaseModel],
+    ]:
+        if get_origin(schema) is TypedDict:
+            raise NotImplementedError("TypedDict is not supported yet by Outlines")
+
+        self.json_schema = schema
+
+        if isinstance(schema, type) and issubclass(schema, BaseModel):
+            parser: Union[PydanticOutputParser, JsonOutputParser] = (
+                PydanticOutputParser(pydantic_object=schema)
+            )
+        else:
+            parser = JsonOutputParser()
+
+        if include_raw:
+            return self | {
+                "raw": RunnablePassthrough(),
+                "parsed": parser,
+                "parsing_error": RunnableLambda(
+                    lambda x: None,  # todo add errors
+                ),
+            }
+
+        return self | parser
+
+    def _generate(
+        self,
+        messages: List[BaseMessage],
+        stop: Optional[List[str]] = None,
+        run_manager: Optional[CallbackManagerForLLMRun] = None,
+        **kwargs: Any,
+    ) -> ChatResult:
+        params = {**self._default_params, **kwargs}
+        if stop:
+            params["stop_at"] = stop
+
+        prompt = self._convert_messages_to_prompt(messages)
+        # todo? stop = return assistant token mask
```

Comment:
could you please turn this into a more descriptive TODO?

Reply:
I think it can be removed was not sure about the stop token but everything works